[Merged by Bors] - Web3Signer support for VC

paulhauner opened the pull request:
[EIP-3030]: https://eips.ethereum.org/EIPS/eip-3030
[Web3Signer]: https://consensys.github.io/web3signer/web3signer-eth2.html

## Issue Addressed

Resolves #2498

## Proposed Changes

Allows the VC to call out to a [Web3Signer] remote signer to obtain signatures.


## Additional Info

### Making Signing Functions `async`

To allow remote signing, I needed to make all the signing functions `async`. This caused a bit of noise where I had to convert iterators into `for` loops.

In `duties_service.rs` there was a particularly tricky case where we couldn't hold a write-lock across an `await`, so I had to first take a read-lock, then grab a write-lock.

### Move Signing from Core Executor

Whilst implementing this feature, I noticed that we signing was happening on the core tokio executor. I suspect this was causing the executor to temporarily lock and occasionally trigger some HTTP timeouts (and potentially SQL pool timeouts, but I can't verify this). Since moving all signing into blocking tokio tasks, I noticed a distinct drop in the "atttestations_http_get" metric on a Prater node:

![http_get_times](https://user-images.githubusercontent.com/6660660/132143737-82fd3836-2e7e-445b-a143-cb347783baad.png)

I think this graph indicates that freeing the core executor allows the VC to operate more smoothly.

### Refactor TaskExecutor

I noticed that the `TaskExecutor::spawn_blocking_handle` function would fail to spawn tasks if it were unable to obtain handles to some metrics (this can happen if the same metric is defined twice). It seemed that a more sensible approach would be to keep spawning tasks, but without metrics. To that end, I refactored the function so that it would still function without metrics. There are no other changes made.

## TODO

- [x] Restructure to support multiple signing methods.
- [x] Add calls to remote signer from VC.
- [x] Documentation
- [x] Test all endpoints
- [x] Test HTTPS certificate
- [x] Allow adding remote signer validators via the API
- [x] Add Altair support via [21.8.1-rc1](https://github.com/ConsenSys/web3signer/releases/tag/21.8.1-rc1)
- [x] Create issue to start using latest version of web3signer. (See #2570)

## Notes

- ~~Web3Signer doesn't yet support the Altair fork for Prater. See https://github.com/ConsenSys/web3signer/issues/423.~~
- ~~There is not yet a release of Web3Signer which supports Altair blocks. See https://github.com/ConsenSys/web3signer/issues/391.~~

paulhauner committed:
Add basic remote signer format

paulhauner committed:
Add support to validator store

paulhauner committed:
Create signing_method.rs

paulhauner committed:
Fix aggregate signing

paulhauner committed:
Add basic request types

paulhauner committed:
Use references for request types

paulhauner committed:
Update validator_store to pre-image

paulhauner committed:
Add default remote signer timeout

paulhauner committed:
Add calls to http server

paulhauner committed:
Refactor task executor metrics

paulhauner committed:
Partially complete async refactor

paulhauner committed:
Refactor aggregator proofs

paulhauner committed:
Remove code comments

paulhauner committed:
Partially address clippy lints

paulhauner committed:
Add signing context

paulhauner committed:
Ignore clippy lint

paulhauner committed:
Rename Web3Signer to remote signer

paulhauner committed:
Fix tests

paulhauner committed:
Add unimplemented Altair block handler

paulhauner committed:
Fix doppelganger check

paulhauner committed:
Add build script for tests

paulhauner committed:
Add tests

paulhauner committed:
Add basic web3 signer rig

paulhauner committed:
Add missed await

paulhauner committed:
Add validator store rig

paulhauner committed:
Add TestingRig

paulhauner committed:
Add first working test

paulhauner committed:
Fix unused vars, tidy

paulhauner committed:
Add block test

paulhauner committed:
Add integration test to CI

paulhauner committed:
Fix integration test

paulhauner committed:
Block test on cargo-fmt

paulhauner committed:
Add more tests, increase upcheck timeout

paulhauner committed:
Add more tests

paulhauner committed:
Fix u64 quoting issue

paulhauner committed:
Separate tests

paulhauner committed:
Add Altair tests for all networks

paulhauner committed:
Fix clippy lints

paulhauner committed:
Move web3signer structs into their own file

paulhauner committed:
Add SignableMessage

paulhauner committed:
Add docs

paulhauner committed:
Add method to API

paulhauner committed:
Add basic test

paulhauner committed:
Rename serde_utils

paulhauner committed:
Add docs for HTTP API

paulhauner committed:
Tidy

paulhauner committed:
Tidy web3signer tests

paulhauner committed:
Tidy, add comments

paulhauner committed:
Use read lock instead of write lock

paulhauner committed:
Tidy

paulhauner committed:
Remove domain mapping

paulhauner committed:
Use with_capacity for vecs

paulhauner committed:
Add signing context function

paulhauner committed:
Allow for fixed web3signer version

paulhauner committed:
Tidy

paulhauner committed:
Add altair block support

paulhauner committed:
Add altair block tests

paulhauner committed:
Use cfg flags to bar windows

paulhauner committed:
Remove repeated web3signer test from CI

paulhauner committed:
Still download artifacts on windows

paulhauner committed:
Use tls in testing

paulhauner committed:
Fix build script

paulhauner committed:
Fix import

paulhauner committed:
Check to see if dir exists in build script

paulhauner committed:
Tidy web3signer types

paulhauner committed:
Use latest release for testing

michaelsproul commented:
I'm reviewing this today

paulhauner committed:
Add comment to build.rs

michaelsproul commented on book/src/validator-web3signer.md:
```diff
@@ -0,0 +1,56 @@
+
+The [Web3Signer] tool is maintained by [Consensys], the same team that maintains [Teku]. The
+Lighthouse team (Sigma Prime) does not maintain Web3Signer or make any guarantees about its safety
+of effectiveness.
```
```suggestion
or effectiveness.
```

michaelsproul commented on book/src/validator-web3signer.md:
```diff
@@ -0,0 +1,56 @@
+"self-signed" SSL certificate from `/home/paul/my-certificates/my-remote-signer.pem` (on the
+filesystem of the VC) to encrypt the communications between the VC and Web3Signer.
+
+> The `request_timeout_ms` key can also be specified. Use this key to override the default time-out
```
```suggestion
> The `request_timeout_ms` key can also be specified. Use this key to override the default timeout
```

michaelsproul commented on book/src/validator-web3signer.md:
```diff
@@ -0,0 +1,56 @@
+filesystem of the VC) to encrypt the communications between the VC and Web3Signer.
+
+> The `request_timeout_ms` key can also be specified. Use this key to override the default time-out
+> with a new timeout in milliseconds. This is the time-out before requests to Web3Signer are
```
```suggestion
> with a new timeout in milliseconds. This is the timeout before requests to Web3Signer are
```

michaelsproul commented on testing/web3signer_tests/src/lib.rs:
```diff
@@ -0,0 +1,575 @@
+                .arg("eth2")
+                .arg(format!("--network={}", network))
+                .arg("--slashing-protection-enabled=false")
+                .spawn()
```
Running these tests produces a lot of output even without `--nocapture`, maybe we should redirect stdout and stderr to `Stdio::null` by default?

michaelsproul commented on validator_client/src/duties_service.rs:
```diff
@@ -651,23 +652,35 @@ async fn poll_beacon_attesters_for_epoch<T: SlotClock + 'static, E: EthSpec>(
-            .map_or(true, |(prior, _)| *prior != dependent_root)
-        {
+        let update_required = {
+            // It is important to ensure that this read-lock is dropped before taking a write-lock
+            // later in the function.
+            let attesters = duties_service.attesters.read();
```
I think we should keep an eye on this locking in case the bouncing causes issues with large numbers of validators. 

One solution would be to switch to Tokio's [`RwLock`](https://docs.rs/tokio/1.11.0/tokio/sync/struct.RwLock.html), but that would require us to `.await` in inconvenient places, like here:

https://github.com/sigp/lighthouse/blob/3e3429754d153e6166c0ff4d0ceb9518799da145/validator_client/src/attestation_service.rs#L188-L197

Alternatively we could do it in 3 steps:

1. Holding the read lock, collect `duty`s that require an update
2. Without any lock held, construct all the `DutyAndProof`s in a vec
3. Holding the write lock, insert the `DutyAndProof`s into the map

paulhauner replied:
Good call. I also noticed that we were exiting early if a single validator failed when I think it's better to just skip it. Especially now signing might involve a HTTP call.

michaelsproul commented on validator_client/src/duties_service/sync.rs:
```diff
@@ -487,77 +489,75 @@ pub fn fill_in_aggregation_proofs<T: SlotClock + 'static, E: EthSpec>(
+                        "Pubkey missing when signing selection proof";
+                        "pubkey" => ?duty.pubkey,
+                        "slot" => slot,
+                    );
```
Would be good to log the error here, seeing as there are now more options than just "key doesn't exist locally"

michaelsproul commented on validator_client/src/signing_method.rs:
```diff
@@ -0,0 +1,215 @@
+
+/// The additional information used to construct a signature. Mostly used for protection from replay
+/// attacks.
+pub struct SigningContext {
+    pub domain: Domain,
+    pub epoch: Epoch,
+    pub fork: Fork,
+    pub genesis_validators_root: Hash256,
+}
```
Ahh this type is so nice :star_struck:

michaelsproul reviewed (changes requested):
Looks great on the whole, I have only a few very minor suggestions!

michaelsproul commented on validator_client/src/sync_committee_service.rs:
```diff
@@ -214,28 +214,30 @@ impl<T: SlotClock + 'static, E: EthSpec> SyncCommitteeService<T, E> {
+                    duty.validator_index,
+                    &duty.pubkey,
+                )
+                .await
```
It just occurred to me that we probably shouldn't be `await`ing these futures in a loop like this.

Unlike the attestation service the sync committee service uses one task for publishing all the signatures. If the VC is managing multiple sync committee members then it will block each sync signature request to the remote signer on the previous one. We could do it in parallel by collecting the futures and joining them.

michaelsproul commented on validator_client/src/sync_committee_service.rs:
```diff
@@ -336,27 +338,29 @@ impl<T: SlotClock + 'static, E: EthSpec> SyncCommitteeService<T, E> {
+                    contribution.clone(),
+                    selection_proof,
+                )
+                .await
```
Similarly there's only one task for signed contributions and proofs, so we should ideally do this in parallel too.

michaelsproul commented on validator_client/src/duties_service/sync.rs:
```diff
@@ -487,77 +489,75 @@ pub fn fill_in_aggregation_proofs<T: SlotClock + 'static, E: EthSpec>(
+                let proof = if let Ok(proof) = duties_service
+                    .validator_store
+                    .produce_sync_selection_proof(&duty.pubkey, slot, subnet_id)
+                    .await
```
These futures should probably also be batched in parallel, as there are potentially quite a lot of them: `num_committee_members * 64 (two epochs lookahead) * 1-4 subnets`. On large networks this is less likely to be an issue, but could be troublesome on testnets with a large fraction of local validators

michaelsproul commented on validator_client/src/attestation_service.rs:
```diff
@@ -377,12 +377,16 @@ impl<T: SlotClock + 'static, E: EthSpec> AttestationService<T, E> {
+                    &mut attestation,
+                    current_epoch,
+                )
+                .await
```
We should probably consider parallelising here too, but it's less urgent because we're unlikely to have multiple attesters for the same committee/index except on small networks.

michaelsproul replied:
There's also an issue whereby the slashing protection database is going to block some amount of parallelism, by requiring that all the slashing protection checks succeed before signing requests are made, see: https://github.com/sigp/lighthouse/issues/1914#issuecomment-917934011

michaelsproul commented on validator_client/src/attestation_service.rs:
```diff
@@ -494,12 +498,16 @@ impl<T: SlotClock + 'static, E: EthSpec> AttestationService<T, E> {
+                    aggregated_attestation.clone(),
+                    selection_proof.clone(),
+                )
+                .await
```
As above.

michaelsproul reviewed (changes requested):
2nd review focusing on performance

paulhauner committed:
Apply suggestions from code review

Co-authored-by: Michael Sproul <micsproul@gmail.com>

paulhauner committed:
Log error

paulhauner committed:
Suppress web3signer logs

paulhauner committed:
Create DutyAndProof in parallel

paulhauner committed:
Perform sync committee tasks in parallel

paulhauner committed:
Perform sync proof signatures in parallel

paulhauner committed:
Produce attestation sigs in parallel

paulhauner committed:
Produce aggregates in parallel

paulhauner committed:
Tidy sync committee service

paulhauner committed:
Tidy sync

paulhauner committed:
Fix web3signer test

paulhauner committed:
Add pubkey to log

paulhauner committed:
Merge branch 'unstable' into remote-signer-client

paulhauner commented:
Thanks for the review @michaelsproul, they were great points about running the tasks in parallel! I've addressed all your comments and this is ready for review again.

You can start your review now, but before we merge I'm going to:

- [x] Run on Prater/Pyrmont.
- [x] Perform manual testing with a testnet validator.

michaelsproul reviewed (approved):
Looks great :ok_hand: 

Merge at will :tada:

paulhauner committed:
Add signing times metrics

paulhauner commented:
Thank you! I add couple more very minor metrics, FYI.

bors r+
